### PR TITLE
[RFC] v_P does not overwrite unnamed registers

### LINKED ISSUE #1790: {visual}P to put without yank to unnamed register
This is one of those things that has always bugged me about vim.  I understand the reason that it's there, but it seems to me as something that behaves the way it does simply due to the way things _used_ to be rather than the way things are now.

What I'm talking about is how 'put' replaces the contents of the default register whenever it's used with any visual block or text object prior to being used.
#### Explanation (for those that need it)

If my text file is:

```
hello
goodbye
```

And I move the cursor to the `goodbye` line and press `yy` to yank it and then move to the `hello` line, visually select and press `p`, if I went to another line to do another put, it wouldn't put `goodbye`, it would put `hello`.

So now if I want to put `goodbye` I have to use the yank register which is two extra keystrokes.  And that only works if I yanked it.  If I deleted it, I have to use a different register.  And now I have to remember if I yanked it or deleted it and use the proper register.
#### End Explanation

I have a mapping to do this, but it barely ever works and is something that IMO needs proper blessed support.

### PR DESCRIPTION
**Pull request by Shougo**
Fix #1790

Continues of #12383

Please test it.

### CODE REVIEW

**Comment by Shougo** [comment]
@jfelchner  Please test it.

Finally I have found the overwrite reason.

**Comment by Shougo** [comment]
`put_spec.lua` is failed.  But it is hard to fix.  It is too long.

**Comment by naruaway** [comment]
Thanks for the PR, I really like the new behavior.
But I think probably we want to keep the existing behavior of `v_p` and just change `v_P` (cf. https://github.com/neovim/neovim/issues/1790#issuecomment-672926687)?

I am not really sure but `v_p` might be used in plugins like `:norm! viwp` and that plugin might be relying on the old behavior. `v_P` is likely not used in plugins since there is no reason to use it instead of `v_p` I believe

**Comment by Shougo** [comment]
OK.  I have only changed `v_P` behavior.

**Comment by Shougo** [comment]
I cannot fix `put_spec.lua` yet.   Please fix it...

File: runtime/doc/vim_diff.txt [hunk 1]
```diff
@@ -360,6 +360,7 @@ Motion:
 Normal commands:
   |Q| replays the last recorded macro instead of switching to Ex mode.
   Instead |gQ| can be used to enter Ex mode.
+  Both |v_P| do not overwrite registers.
```
Comment (zeertzjq) [inline runtime/doc/vim_diff.txt:363]:
```suggestion
  |v_P| does not overwrite registers.
```

File: src/nvim/normal.c [hunk 2]
```diff
@@ -7490,9 +7490,9 @@ static void nv_put_opt(cmdarg_T *cap, bool fix_indent)
       // do_put(), which requires the visual selection to still be active.
       if (!VIsual_active || VIsual_mode == 'V' || regname != '.') {
         // Now delete the selected text. Avoid messages here.
+        cap->oap->regname = cap->cmdchar == 'P' ? '_' : NULL;
```
Comment (zeertzjq) [inline src/nvim/normal.c:7493]: `regname` is not a pointer so it should be `NUL`.
```suggestion
        cap->oap->regname = cap->cmdchar == 'P' ? '_' : NUL;
```

Comment (Shougo) [reply src/nvim/normal.c]: Oh, it is my fault.

**Comment by zeertzjq** [comment]
This fixes tests:
```diff
diff --git a/test/functional/editor/put_spec.lua b/test/functional/editor/put_spec.lua
index 26967ecbb..234579dce 100644
--- a/test/functional/editor/put_spec.lua
+++ b/test/functional/editor/put_spec.lua
@@ -507,7 +507,9 @@ describe('put command', function()
       return function(exception_table, after_redo)
         test_expect(exception_table, after_redo)
         if selection_string then
-          eq(getreg('"'), selection_string)
+          if not conversion_table.put_backwards then
+            eq(getreg('"'), selection_string)
+          end 
         else
           eq(getreg('"'), 'test_string"')
         end
@@ -714,7 +716,9 @@ describe('put command', function()
             expect_base, conversion_table)
           return function(exception_table, after_redo)
             test_expect(exception_table, after_redo)
-            eq(getreg('"'), 'Line of words 1\n')
+            if not conversion_table.put_backwards then
+              eq(getreg('"'), 'Line of words 1\n')
+            end
           end
         end
         local base_expect_string = [[
@@ -748,7 +752,9 @@ describe('put command', function()
           end, expect_base, conversion_table)
         return function(e,c)
           test_expect(e,c)
-          eq(getreg('"'), 'Lin\nLin')
+          if not conversion_table.put_backwards then
+            eq(getreg('"'), 'Lin\nLin')
+          end
         end
       end
```

The `eq()` is used very strangely in the tests. I'll open a PR to fix it.

**Comment by Shougo** [comment]
Thank you!  It fix the problem.

**Comment by zeertzjq** [comment]
#17134 was merged, so this now has conflicts and needs a rebase.

**Comment by Shougo** [comment]
I have fixed the conflicts.

**Comment by Shougo** [comment]
I may port the patch to Vim.

**Comment by Shougo** [comment]
I have ported it to Vim.
https://github.com/vim/vim/pull/9591

**Comment by Shougo** [comment]
well, neovim has not `get_y_current()`, `save_y_current()`.
It is hard to port...

**Comment by Shougo** [comment]
I think `op_reg_set_previous()` is same function.   But I don't know how to get it.

**Pull request closed without merging**